internal/pkg/configmanager/manager.go

```diff
@@ -0,0 +1,271 @@
+		log.Errorf("Please verify the format of your config. Error: %s.", err)
+		return nil, err
+	}
+
+	// 3. render ci/cd templates
+	cis := make([][]PipelineTemplate, 0)
+	cds := make([][]PipelineTemplate, 0)
+	for i, app := range config.AppsInConfig {
+		cisOneApp, err := renderCICDFromPipeTemplates(app.CIs, config.PipelineTemplates, globalVars, i, app.Name, "ci")
+		if err != nil {
+			return nil, err
+		}
+		cis = append(cis, cisOneApp)
+
+		cdsOneApp, err := renderCICDFromPipeTemplates(app.CDs, config.PipelineTemplates, globalVars, i, app.Name, "cd")
+		if err != nil {
+			return nil, err
+		}
+		cds = append(cds, cdsOneApp)
+	}
+
+	// remove the pipeline templates, beacuse we don't need them anymore.
+	// and because vars here main contains local vars,
+	// it will cause error when rendered in the next step if we don't remove them.
+	config.PipelineTemplates = nil
+
+	// 4. re-generate the config bytes(without pipeline templates)
+	configBytesWithoutPipelineTemplates, err := yaml.Marshal(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// 5. render all vars to the whole config bytes
+	renderedConfigBytes, err := renderConfigWithVariables(string(configBytesWithoutPipelineTemplates), globalVars)
+	if err != nil {
+		return nil, err
+	}
+
+	renderedConfigStr := string(renderedConfigBytes)
+	log.Debugf("redenered config: %s\n", renderedConfigStr)
+
+	// 6. yaml unmarshal again to get the whole config
+	var configRendered ConfigRaw
+	//renderedConfigStr := string(renderedConfigBytes)
+	//fmt.Println(renderedConfigStr)
+	err = yaml.Unmarshal(renderedConfigBytes, &configRendered)
+	if err != nil {
+		return nil, err
+	}
+	configFinal := &Config{}
+	configFinal.PluginDir = configRendered.PluginDir
+	configFinal.State = configRendered.State
+	configFinal.Tools = configRendered.Tools
+
+	// 7. restructure the apps
+	for i, app := range configRendered.AppsInConfig {
+		appFinal := App{
+			Name:         app.Name,
+			Spec:         app.Spec,
+			Repo:         app.Repo,
+			RepoTemplate: app.RepoTemplate,
+		}
+		appFinal.CIs = cis[i]
+		appFinal.CDs = cds[i]
+		configFinal.Apps = append(configFinal.Apps, appFinal)
+	}
+
+	errs := configFinal.Validate()
+	if len(errs) > 0 {
+		return nil, multierr.Combine(errs...)
+	}
+
+	return configFinal, nil
+}
+
+func getVarsFromConfigBytes(configBytes []byte) (map[string]any, error) {
+	// how to get all the global vars:
+	// we regard the whole config bytes as a map,
+	// and we regard all the key-value as global vars(except some special keys)
+	allVars := make(map[string]any)
+	if err := yaml.Unmarshal(configBytes, allVars); err != nil {
+		return nil, err
+	}
+
+	excludeKeys := []string{
+		"varFile", "toolFile", "appFile", "templateFile", "pluginDir",
+		"state", "tools", "apps", "pipelineTemplates",
+	}
+
+	for _, key := range excludeKeys {
+		// delete the special keys
+		delete(allVars, key)
+	}
+
+	return allVars, nil
+}
+
+func mergeMaps(m1 map[string]any, m2 map[string]any) map[string]any {
```

Comment:
If map1 and map2 share a common key, the value would come from map2. I'm not sure if this should be the behaviour because if a user accidentally defines the same var in two maps, it'd be hard to debug.

Reply:
Also I think it'd be nice if we could have some UT covering this basic function.